### PR TITLE
Player control

### LINKED ISSUE #3: Player control
Player controls:
- walk
- run
- simple jump with height affected by how long the player press jump
- sliding on walls
- jumping from walls

### PR DESCRIPTION
**Pull request by tienlocnguyen**
Main purpose of this PR is control player. I also implement simple scene to test it. Test scene can be used for action level, but it simple and quality is not really good.

### CODE REVIEW

**Review by T4g1 — CHANGES_REQUESTED** [review]
- Please remove warnings and debug printing (12 warnings and many more debug print)
- Some inconsistency when moving player: If I press left then right and stay on those: Player goes left. If I press right then left, player goes left (behavior expected is to stay going right then): I suggest to stop when both key are pressed.
- Pressing up once will make the player jump if held too much: Can trigger unwanted jumps

**Comment by T4g1** [comment]
Also the test scene should probably not be part of it right now (we don't want debug things on master?)

**Comment by tienlocnguyen** [comment]
> Also the test scene should probably not be part of it right now (we don't want debug things on master?)
Yes, it's for test only. I'll remove it.

**Comment by tienlocnguyen** [comment]
> * much

I fixed point 1 and 2. Still have workaround to fix warning in signal handle(some param). It will be updated later when have enough game entity like obstacles
Point 3 is not clear. I hold up button but there's no unwanted jump. In my code, player only jump when he's not jumping

File: project.godot [hunk 1]
```diff
@@ -8,16 +8,29 @@
...
 config/name="gd-wildjam18"
 config/icon="res://icon.png"
 
+[layer_names]
+
+2d_physics/layer_1="Floor"
+2d_physics/layer_2="Pipe"
+2d_physics/layer_3="DeathHole"
+2d_physics/layer_11="PlayerBody"
+2d_physics/layer_12="PayerStandShape"
```
Comment (T4g1) [inline project.godot]: Why do we need two layers for the player?
What is a Pipe? Why is an "obstacle" layer not enough for Pipe and DeathHole?
What is DeathHole?

Comment (T4g1) [reply project.godot]: What I believe should be enough is:
Player, Obstacle, Trigger (or Area)

File: sources/player/Player.tscn [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+[sub_resource type="RectangleShape2D" id=4]
+extents = Vector2( 30, 30 )
+
+[node name="Player" type="KinematicBody2D"]
+collision_layer = 1024
+collision_mask = 7
+script = ExtResource( 2 )
+
+[node name="PlayerShape" type="CollisionShape2D" parent="."]
+position = Vector2( 150, 97 )
```
Comment (T4g1) [inline sources/player/Player.tscn]: Typo? You probably want player to be centered by default with no offset (0; 0)

File: sources/player/Player.tscn [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+shape = SubResource( 1 )
+
+[node name="Sprite" type="Sprite" parent="."]
+position = Vector2( 150, 97 )
```
Comment (T4g1) [inline sources/player/Player.tscn]: Same here

File: sources/player/Player.tscn [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+texture = ExtResource( 1 )
+
+[node name="AnimationPlayer" type="AnimationPlayer" parent="."]
+anims/Crouch = SubResource( 2 )
+anims/Standup = SubResource( 3 )
+
+[node name="StandShape" type="Area2D" parent="."]
+position = Vector2( 150, 96 )
```
Comment (T4g1) [inline sources/player/Player.tscn]: And here

**Review by T4g1 — CHANGES_REQUESTED** [review]
About the jumping logic: Wouldn't it be better to push the up button to stand and not jump if we hold it? When pressing the first time, player want to stand, not jump, thats why I believe, jumping after a delay makes it inconsistent with player expectation

About the crouching: Is it necessary to hold the crouching stance when crouch is not held? As we are designing a running/action scene and not a shooter, I believe Player need to stand up as fast as possible  (implying speed is reduced in crouching stance)

About DeathHole: Wouldn't it be better to separate the DeathHole logic from the player?

Side note: You can use https://docs.godotengine.org/en/3.2/classes/class_area2d.html#class-area2d-method-get-overlapping-bodies with StandShape instead of relying on multiple signals and counter

Also for simplification, I recommend you this: https://docs.godotengine.org/en/3.2/tutorials/2d/2d_movement.html#way-movement Where they separate input logic and movement logic (right now, the whole movement/jump/crouch/input logic is closely interwinned which makes it difficult to understand)

File: sources/player/Player.tscn (hunk 2 above)
Comment (tienlocnguyen) [reply sources/player/Player.tscn]: I used editor and move player around. So I don't note this change. Will update it

File: sources/player/Player.tscn (hunk 3 above)
Comment (tienlocnguyen) [reply sources/player/Player.tscn]: Same as above

File: sources/player/Player.tscn (hunk 4 above)
Comment (tienlocnguyen) [reply sources/player/Player.tscn]: Same as above

File: project.godot (hunk 1 above)
Comment (tienlocnguyen) [reply project.godot]: Sorry for my English. I lacked some vocabularies.
About 2 layers for players: I need 2 collide shapes for players:
- One is player shape which change when he stand, crouch
- One is shape when he stand. I used it to make sure player can stand while he is crouching. For example, when he enter a location that not high enough to stand but enough to crouch, he can't stand any more
image
Pipe is think that player can jump up to run
DeathHole: When player collide it, player die. My idea is a death hole will move after player, so player need move continuously.

For player movement, I base on demo in: https://github.com/godotengine/godot-demo-projects/blob/master/2d/kinematic_character/player.gd. Seem that it have some update compare with my local version

Comment (T4g1) [reply project.godot]: Yeah but the point is: Why not put Pipe in an Obstacle layer (along with floor and so on). Standing shape and Death Hole in a Trigger layer? And Player body in the Player layer?

File: project.godot [hunk 5]
```diff
@@ -45,36 +50,21 @@ _global_script_class_icons={
...
 [layer_names]
 
-2d_physics/layer_1="player"
-2d_physics/layer_2="interactable"
-2d_physics/layer_3="obstacle"
-
-[locale]
-
-translations=PoolStringArray( "res://assets/translation/text.en.translation", "res://assets/translation/text.es.translation" )
+2d_physics/layer_1="Platform"
```
Comment (T4g1) [inline project.godot]: Renaming the layers already defined is probably not what you wanted to do here. Use layer after the third one if you need more (but again, do we need a layer for Wall and Platform? Obstacle isn't enough? Is this relevant to player movement?

File: sources/player/Player.gd [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	var force = Vector2(0, Gravity)
+
+	var walk = Input.get_action_strength("ui_right") - Input.get_action_strength("ui_left")
+	var jump = Input.is_action_pressed("ui_up")
+	
+	# while jumping, player can't change direction. Can be update later
+	if not jumping:
+		force.x = update_horizontal_force(walk, delta)
+		
+	velocity += force * delta
```
Comment (T4g1) [inline sources/player/Player.gd]: Regarding move_and_slide: https://docs.godotengine.org/en/3.1/classes/class_kinematicbody2d.html#class-kinematicbody2d-method-move-and-slide You should not multiply it by delta, its already taken care of.

File: sources/player/Player.tscn (hunk 2 above)
Comment (T4g1) [reply sources/player/Player.tscn]: Still relevant here

File: sources/player/Player.tscn (hunk 3 above)
Comment (T4g1) [reply sources/player/Player.tscn]: And here

**Review by T4g1 — CHANGES_REQUESTED** [review]
Way better, I commented with what cannot go on master and here are some general tips: Its somewhat unclear what most of the function do. Adding a few comments could help. For example, do we want a function to jump the player or a function to try_to_jump?

Also, the coding conventions are inconsistent: Naming is sometimes camelcase, uppercase, short, long. What is a vlen? or a vsign? Why not use velocity_sign for example (if it's indeed what you mean by vsign)
There is a method named my_method and a variable named on_air_time which is inconsistent too

**Review by T4g1 — APPROVED** [review]

**Pull request merged**